native/xqlitenif/src/r2d2.rs

```diff
@@ -0,0 +1,445 @@
+            XqliteError::CannotReadColumn(index, reason) => {
+                (cannot_read_column(), index, reason).encode(env)
+            }
+            XqliteError::CannotExecute(reason) => (cannot_execute(), reason).encode(env),
+            XqliteError::CannotExecutePragma { pragma, reason } => {
+                (cannot_execute_pragma(), pragma, reason).encode(env)
+            }
+            XqliteError::CannotFetchRow(reason) => (cannot_fetch_row(), reason).encode(env),
+            XqliteError::CannotAllocateBinary => cannot_allocate_binary().encode(env),
+            XqliteError::CannotOpenDatabase(path, reason) => {
+                (cannot_open_database(), path, reason).encode(env)
+            }
+            XqliteError::CannotConvertAtomToString(reason) => {
+                (cannot_convert_atom_to_string(), reason).encode(env)
+            }
+        }
+    }
+}
+
+impl RefUnwindSafe for XqliteError {}
+
+impl From<r2d2::Error> for XqliteError {
+    fn from(err: r2d2::Error) -> Self {
+        XqliteError::Timeout(err.to_string())
+    }
+}
+
+impl From<rusqlite::Error> for XqliteError {
+    fn from(err: rusqlite::Error) -> Self {
+        XqliteError::CannotExecute(err.to_string())
+    }
+}
+
+fn elixir_term_to_rusqlite_value<'a>(
+    env: Env<'a>,
+    term: Term<'a>,
+) -> Result<Value, XqliteError> {
+    let make_convert_error = |term: Term<'a>, err: RustlerError| -> XqliteError {
+        XqliteError::CannotConvertToSqliteValue {
+            value_str: format!("{:?}", term),
+            reason: format!("{:?}", err),
+        }
+    };
+
+    let term_type = term.get_type();
+    match term_type {
+        TermType::Atom => {
+            if term == nil().to_term(env) {
+                Ok(Value::Null)
+            } else if term == r#true().to_term(env) {
+                Ok(Value::Integer(1))
+            } else if term == r#false().to_term(env) {
+                Ok(Value::Integer(0))
+            } else {
+                Err(XqliteError::UnsupportedAtom {
+                    atom_value: term
+                        .atom_to_string()
+                        .unwrap_or_else(|_| format!("{:?}", term)),
+                })
+            }
+        }
+        TermType::Integer => term
+            .decode::<i64>()
+            .map(Value::Integer)
+            .map_err(|e| make_convert_error(term, e)),
+        TermType::Float => term
+            .decode::<f64>()
+            .map(Value::Real)
+            .map_err(|e| make_convert_error(term, e)),
+        TermType::Binary => match term.decode::<String>() {
+            Ok(s) => Ok(Value::Text(s)),
+            Err(_string_decode_err) => match term.decode::<Binary>() {
+                Ok(bin) => Ok(Value::Blob(bin.as_slice().to_vec())),
+                Err(binary_decode_err) => Err(make_convert_error(term, binary_decode_err)),
+            },
+        },
+        _ => Err(XqliteError::UnsupportedDataType { term_type }),
+    }
+}
+
+fn decode_keyword_params<'a>(
```

Comment:
I'd personally not do this in Rust but in Elixir instead, only passing "pre-checked" maps into the functions that are unpacked using [`NifMap`](https://docs.rs/rustler/latest/rustler/derive.NifMap.html) types.

Reply:
We need to parse the Elixir terms into Rust objects anyway (for parameterized queries) -- that is unavoidable, so why not go all the way and validate it properly as we go? Otherwise we would duplicate work if we f.ex. use `NimbleOptions` on the Elixir side. Am I misunderstanding you?

Reply:
Forgot to clarify: I really would prefer keyword lists for parameterized SQL queries. They just read and are written better in Elixir.

Though thinking of it, maybe you are still right: I can make the Rust code only use the pre-checked `NifMap` types and `raw_exec` should only work with them -- but then have higher-level Elixir helpers that accepts keyword lists and convert them to maps before passing them to `raw_exec`. Is that how would you go about it?

Reply:
Cool, I'll give it a shot after we figure out the zero-copy thing.

I wonder if such a function might still not be a valuable addition to Rustler though. 🤔

Reply:
Thought long and hard about this but ultimately decided that 100-200 extra Rust lines are not bothering me. I prefer to centralize the checks and not rely on Elixir code wrappers. Sure, if the users of the library opt for "raw" functions and pass wrong values and get a crash it's definitely on them and it's not something I should optimize for... but I simply want to have the absolute minimum of Elixir wrappers.

There's no right or wrong answer here, I feel. Your take is just as valid. But ultimately that was my preference: centralize those checks inside the Rust code.